common/appscale/common/async_retrying.py

```diff
@@ -0,0 +1,305 @@
+      backoff_threshold: a number indicating maximum backoff.
+      max_retries: an integer indicating maximum number of retries.
+      retrying_timeout: a number indicating number of seconds after which
+        retrying should be stopped.
+      retry_on_exception: a function receiving one argument: exception object
+        and returning True if retry makes sense for such exception.
+        Alternatively you can pass list of exception types for which
+        retry is needed.
+    """
+    self.backoff_base = not_missed(backoff_base, DEFAULT_BACKOFF_BASE)
+    self.backoff_multiplier = not_missed(backoff_multiplier,
+                                         DEFAULT_BACKOFF_MULTIPLIER)
+    self.backoff_threshold = not_missed(backoff_threshold,
+                                        DEFAULT_BACKOFF_THRESHOLD)
+    self.max_retries = not_missed(max_retries, DEFAULT_MAX_RETRIES)
+    self.retrying_timeout = not_missed(retrying_timeout,
+                                       DEFAULT_RETRYING_TIMEOUT)
+    self.retry_on_exception = not_missed(retry_on_exception,
+                                         DEFAULT_RETRY_ON_EXCEPTION)
+
+  def __call__(self, func=None, backoff_base=MISSED, backoff_multiplier=MISSED,
+               backoff_threshold=MISSED, max_retries=MISSED,
+               retrying_timeout=MISSED, retry_on_exception=MISSED):
+    """ Wraps python generator with tornado coroutine and retry mechanism
+    which runs up to max_retries attempts with exponential backoff
+    (sleep = backoff_multiplier * backoff_base**X).
+
+    Args:
+      func: a function to wrap.
+      backoff_base: a number to use in backoff calculation.
+      backoff_multiplier: a number indicating initial backoff.
+      backoff_threshold: a number indicating maximum backoff.
+      max_retries: an integer indicating maximum number of retries.
+      retrying_timeout: a number indicating number of seconds after which
+        retrying should be stopped.
+      retry_on_exception: a function receiving one argument: exception object
+        and returning True if retry makes sense for such exception.
+        Alternatively you can pass list of exception types for which
+        retry is needed.
+    Returns:
+      A wrapped coroutine or self if coroutine arg was omitted in arguments.
+    """
+    if func is None:
+      return _RetryCoroutine(
+        backoff_base=backoff_base,
+        backoff_multiplier=backoff_multiplier,
+        backoff_threshold=backoff_threshold,
+        max_retries=max_retries,
+        retrying_timeout=retrying_timeout,
+        retry_on_exception=retry_on_exception
+      )
+
+    args = (backoff_base, backoff_multiplier, backoff_threshold, max_retries,
+            retrying_timeout, retry_on_exception)
+    if any((argument is not MISSED) for argument in args):
+      raise TypeError('You should pass either func or parameters')
+
+    coroutine = gen.coroutine(func)
+
+    @functools.wraps(coroutine)
```

Comment:
I'm having trouble understanding why this would wrap coroutine rather than func.